enhancements/update/cvo-log-level-api.md

```diff
@@ -0,0 +1,325 @@
+---
+title: cvo-log-level-api
+authors:
+  - "@Davoska"
+reviewers:
+  - "@wking"
+  - "@LalatenduMohanty"
+  - "@petr-muller"
+approvers:
+  - "@wking"
+api-approvers:
+  - "@deads2k"
+  - "@bparees"
+creation-date: 2023-10-09
+last-updated: 2024-09-10
+tracking-link:
+  - https://issues.redhat.com/browse/OTA-923
+see-also:
+  - "None"
+replaces:
+  - "None"
+superseded-by:
+  - "None"
+---
+
+# CVO Log Level API
+
+## Summary
+
+This enhancement describes the API changes needed to provide a simple
+way of dynamically changing the verbosity level of Cluster Version Operator's logs.
+
+## Motivation
+
+The [Cluster Version Operator (CVO)](https://github.com/openshift/cluster-version-operator)
+is responsible for a number of resources and thus logs a lot of information.
+However, there is currently no way to easily change the verbosity of the CVO logs.
+
+It would be useful to provide functionality for the cluster administrators and
+OpenShift engineers to easily modify the log level to a desired level similarly
+as can be done for [`ClusterOperators`](https://github.com/openshift/api/blob/5852b58f4b1071fe85d9c49dff2667a9b2a74841/operator/v1/types.go#L58-L65).
+
+### User Stories
+
+* As an OpenShift administrator, I want to increase the log level of the CVO to
+  more easily troubleshoot any potential issues regarding the cluster or CVO.
+* As an OpenShift administrator, I want to decrease the log level of the CVO to
+  save up more storage space.
+* As an OpenShift engineer, I want to increase the log level of the CVO for
+  the CI runs so that I can more easily troubleshoot any potential issues that occurred.
+* As an OpenShift engineer, I want to decrease the log level of the CVO for
+  shipped releases so that customers don't receive debug logs.
+
+### Goals
+
+The goal is to add a user-facing API for controlling the verbosity of the CVO logs.
+
+### Non-Goals
+
+Change the default logging verbosity of the Cluster Version Operator in production
+OCP clusters.
+
+## Proposal
+
+This enhancement proposes to add a new field to the `ClusterVersion` Custom Resource Definition (CRD). The CVO
```

Comment:
but we can't do this because of the import cycles discussed in the implementation section.  Well, we could give the CVO it's own `OperatorSpec` consumer, but it seems strange to have the CVO configured by ClusterVersion.config.openshift.io and also ClusterVersionOperator.operator.openshift.io.  Is there a UX benefit to that kind of sharding?  At the moment, we're lifting this config into ClusterVersion, because we think one-CVO-config-resource is the more usable UX, even if it's a bit annoying from an API-maintainer perspective.

Reply:
I'm actually of fan of splitting cluster and cvo config. ClusterVersion would not configure CVO but the cluster. New resource would configure CVO. ClusterVersion should tell what CVO should do. New resource would tweak how.

Reply:
> ClusterVersion would not configure CVO but the cluster. New resource would configure CVO. 

^^^ this

The typical split is along an operator/operand boundary. The operand for the CVO is the cluster, so the existing configuration represents what the state of the cluster should be. If you want to represent how the operator behaves, then that's an operator group CRD if you look at how other operators have built out their configuration.

Reply:
Yes, I have spent some time learning more about the OpenShift API and have discussed the possible options of this enhancement with the team. We have decided to propose the creation of a new ClusterVersionOperator CRD as suggested by you, Joel. I am reworking this enhancement at the moment. ﻿

Reply:
A new `ClusterVersionOperator` CRD is proposed in the changes.